File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.cpp
#include <vector>
#include <stack>
using namespace std;

vector<int> solution(vector<int> numbers) {
    vector<int> answer(numbers.size(), -1);
    stack<int> s;
    for (int i = 0; i < numbers.size(); i++) {
        int n = numbers[i];
        while (!s.empty() && numbers[s.top()] < n) {
            answer[s.top()] = n;
            s.pop();
        }
        s.push(i);        
    }
    return answer;
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.length];
        Arrays.fill(answer, -1);
        
        Deque<Integer> stk = new ArrayDeque<>();
        stk.push(0);
        
        for (int i = 1; i < numbers.length; i++) {
            while (!stk.isEmpty() && numbers[stk.peek()] < numbers[i]) {
                int idx = stk.pop();
                answer[idx] = numbers[i];
            }
            stk.push(i);
        }
    
        return answer;
    }
}